### PR TITLE
New Pallet: Root offences

### LINKED ISSUE #11352: Allow submitting offences from root
Would be very helpful for testing if there was a way for `root` e.g. sudo in a dev-chain to create an `offence`. This would help you test all sorts of things around the slashing and offence code. 

Ideally, this transaction would only accept one parameters: `offenders: Vec<(T::AccountId, Perbill)>`, and slash each of them with that fraction. 

One way to do this would be to add this transaction to the offences pallet. But it won't be able to work with the api above (offences pallet needs `FullIdentification`), and it would also pollute all runtime with this pallet.

An alternative is to build a custom, minimal pallet that only does this. 

This new pallet should strictly depend on `pallet-session`, `pallet-staking`, and `pallet-offences` in its `trait Config`.

Having all 3 of the above, realizing the interface `offenders: Vec<(T::AccountId, Perbill)>` should also be possible.

### PR DESCRIPTION
**Pull request by Szegoo**
Closes #11352

Polkadot address: 126X27SbhrV19mBFawys3ovkyBS87SGfYwtwa8J2FjHrtbmA

### CODE REVIEW

**Comment by Szegoo** [comment]
@kianenigma Could you take a look, and check if I am on the right path with this? I am not done here, but before I continue I just want to make sure that I am not doing something wrong here.

File: frame/system/src/lib.rs [hunk 1]
```diff
@@ -278,7 +278,8 @@ pub mod pallet {
 			+ Debug
 			+ MaybeDisplay
 			+ Ord
-			+ MaxEncodedLen;
+			+ MaxEncodedLen
+			+ Clone;
```
Comment (KiChjang) [inline frame/system/src/lib.rs:282]: This is unnecessary, `Parameter` already implies `Clone`: https://paritytech.github.io/substrate/master/frame_support/dispatch/trait.Parameter.html

Comment (Szegoo) [reply frame/system/src/lib.rs]: Yeah, this didn't make much sense, other than that does the logic seem ok?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1745227

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1755482

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1755484

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs. 
 The job name - test-linux-stable 
 The job logs - https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/1755532

File: frame/nomination-pools/src/lib.rs [hunk 2]
```diff
@@ -2479,8 +2479,8 @@ impl<T: Config> Pallet<T> {
 impl<T: Config> OnStakerSlash<T::AccountId, BalanceOf<T>> for Pallet<T> {
 	fn on_slash(
 		pool_account: &T::AccountId,
-		// Bonded balance is always read directly from staking, therefore we need not update
-		// anything here.
+		// Bonded balance is always read directly from staking, therefore we
```
Comment (kianenigma) [inline frame/nomination-pools/src/lib.rs:2482]: please make sure lines are wrapped at 100, not less.

File: frame/root-offences/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+try-runtime = []
+default = ["std"]
+std = [
+	"codec/std",
+	"scale-info/std",
+	"pallet-session/std",
+	"pallet-staking/std",
+	"pallet-offences/std",
+	"frame-support/std",
+    "frame-system/std",
```
Comment (kianenigma) [inline frame/root-offences/Cargo.toml:47]: bad indentation.

File: frame/root-offences/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+pallet-timestamp = { version = "4.0.0-dev", path = "../timestamp" }
+pallet-staking-reward-curve = { version = "4.0.0-dev", path = "../staking/reward-curve" }
+
+sp-core = { version = "6.0.0", path = "../../primitives/core" }
+sp-io = { version = "6.0.0", default-features = false, path = "../../primitives/io" }
+frame-election-provider-support = { version = "4.0.0-dev", path = "../election-provider-support" }
+
+[features]
+runtime-benchmarks = []
+try-runtime = []
```
Comment (kianenigma) [inline frame/root-offences/Cargo.toml:38]: should enable `frame-support/try-runtime` and a few others, look into other pallets.

File: frame/root-offences/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	pub enum Event<T: Config> {
+		RootCreatedOffence { offenders: Vec<(T::AccountId, Perbill)> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		FailedToGetActiveEra,
+	}
+
+	#[allow(type_alias_bounds)]
```
Comment (kianenigma) [inline frame/root-offences/src/lib.rs:63]: why is this needed?

Comment (kianenigma) [reply frame/root-offences/src/lib.rs:63]: You should be able to do `<T as Config>::Stuff` instead and not bound the `<T: Config>`.

File: frame/root-offences/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+	type OffenceDetails<T: Config> =
+		sp_staking::offence::OffenceDetails<T::AccountId, IdentificationTuple<T>>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T>
+	where
+		T: pallet_session::Config<ValidatorId = <T as frame_system::Config>::AccountId>,
```
Comment (kianenigma) [inline frame/root-offences/src/lib.rs:70]: wouldn't it be better to ensure all of these bounds when defining `trait Config`?

File: frame/root-offences/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+			origin: OriginFor<T>,
+			offenders: Vec<(T::AccountId, Perbill)>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let slash_fraction = offenders
+				.clone()
+				.into_iter()
+				.map(|(_, fraction)| fraction)
+				.collect::<Vec<Perbill>>();
```
Comment (kianenigma) [inline frame/root-offences/src/lib.rs:92]:
```suggestion
				.collect::<Vec<_>>();
```

File: frame/root-offences/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+				Weight,
+			>>::on_offence(&offenders, &slash_fraction, session_index, DisableStrategy::WhenSlashed);
+		}
+
+		/// Returns a vector of offenders that are going to be slashed.
+		fn get_offence_details(
+			offenders: Vec<(T::AccountId, Perbill)>,
+		) -> Result<Vec<OffenceDetails<T>>, DispatchError> {
+			let active_era = Staking::<T>::active_era().ok_or(Error::<T>::FailedToGetActiveEra)?;
+			let now = active_era.index;
```
Comment (kianenigma) [inline frame/root-offences/src/lib.rs:127]:
```suggestion
			let now = Staking::<T>::active_era().map(|e| e.index).ok_or(Error::<T>::FailedToGetActiveEra)?;
```

**Review by kianenigma — COMMENTED** [review]
Minor complains, but overall looks good.

File: frame/nomination-pools/src/lib.rs (hunk 2 above)
Comment (Szegoo) [reply frame/nomination-pools/src/lib.rs:2482]: I use rewrap for wrapping all of my comments as you recommended, not sure why it sometimes fails to wrap at 100.

File: frame/root-offences/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	type OffenceDetails<T> = sp_staking::offence::OffenceDetails<
+		<T as frame_system::Config>::AccountId,
+		IdentificationTuple<T>,
+	>;
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Allows the `root`, for example sudo to create an offence.
+		#[pallet::weight(T::DbWeight::get().reads(2))]
```
Comment (Szegoo) [inline frame/root-offences/src/lib.rs:81]: @kianenigma Should I write a benchmark for this or would it be possible somehow to add the weight consumed when calling the `on_offence` function to this?

**Comment by Szegoo** [comment]
@KiChjang @kianenigma Could you please review this?

File: frame/root-offences/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		CreatedOffence { offenders: Vec<(T::AccountId, Perbill)> },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		FailedToGetActiveEra,
```
Comment (kianenigma) [inline frame/root-offences/src/lib.rs:71]: Some description would be good.

File: frame/root-offences/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+	}
+
+	type OffenceDetails<T> = sp_staking::offence::OffenceDetails<
+		<T as frame_system::Config>::AccountId,
+		IdentificationTuple<T>,
+	>;
```
Comment (kianenigma) [inline frame/root-offences/src/lib.rs:77]:
```suggestion
	>;

```

File: frame/root-offences/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+			ensure_root(origin)?;
+
+			let slash_fraction =
+				offenders.clone().into_iter().map(|(_, fraction)| fraction).collect::<Vec<_>>();
+
+			let offence_details = Self::get_offence_details(offenders.clone())?;
+
+			Self::submit_offence(&offence_details, &slash_fraction);
+
+			Self::deposit_event(Event::CreatedOffence { offenders });
```
Comment (kianenigma) [inline frame/root-offences/src/lib.rs:95]:
```suggestion
			let slash_fraction =
				offenders.clone().into_iter().map(|(_, fraction)| fraction).collect::<Vec<_>>();
			let offence_details = Self::get_offence_details(offenders.clone())?;

			Self::submit_offence(&offence_details, &slash_fraction);
			Self::deposit_event(Event::CreatedOffence { offenders });
```

**Review by kianenigma — APPROVED** [review]
LGTM, thanks!

**Comment by Szegoo** [comment]
> /tip large

@kianenigma Hmm, this didn't work :/

File: frame/root-offences/README.md [hunk 13]
```diff
@@ -0,0 +1,5 @@
+# Sudo Offences Pallet
+
+Pallet that allows the root to create an offence.
+
+NOTE: This pallet should be used for testing purposes.
```
Comment (ordian) [inline frame/root-offences/README.md:5]:
```suggestion
NOTE: This pallet should only be used for testing purposes.
```

File: frame/root-offences/src/mock.rs [hunk 14]
```diff
@@ -0,0 +1,339 @@
+use super::*;
```
Comment (ordian) [inline frame/root-offences/src/mock.rs:1]: nit: missing license header

File: frame/root-offences/src/tests.rs [hunk 15]
```diff
@@ -0,0 +1,77 @@
+use super::*;
```
Comment (ordian) [inline frame/root-offences/src/tests.rs:1]: nit: missing license header

File: frame/root-offences/Cargo.toml [hunk 16]
```diff
@@ -0,0 +1,51 @@
+[package]
+name = "pallet-root-offences"
+version = "1.0.0"
+authors = ["Parity Technologies <admin@parity.io>"]
+edition = "2021"
+license = "Apache-2.0"
+homepage = "https://substrate.io"
+repository = "https://github.com/paritytech/substrate/"
+description = "FRAME sudo offences pallet"
```
Comment (ordian) [inline frame/root-offences/Cargo.toml:9]:
```suggestion
description = "FRAME root offences pallet"
```

**Review by ordian — APPROVED** [review]

**Comment by Szegoo** [comment]
@kianenigma I think we can merge this now :)

**Comment by ordian** [comment]
bot merge

**Comment by kianenigma** [comment]
\tip large

**Comment by kianenigma** [comment]
@Szegoo the tip bot is currently down. If I forget, please remind me again and I will make sure it happens. You can reach out to @kianenigma:matrix.parity.io or kian@parity.io.

**Comment by shawntabrizi** [comment]
\tip large

**Comment by Szegoo** [comment]
Thanks!
